### PR TITLE
Update Mosviz parser to load level 2 data

### PR DESCRIPTION
**Pull request by rosteen**
Does what it says on the tin. In-progress, currently (mostly) working for NIRSpec level 2.

EDIT: ready for review, now also loads NIRCam data. I have NIRCam test data, I can upload that to Box (I trimmed a couple real files from ~1200 sources down to 20 to speed things up) and add a test.

### CODE REVIEW

**Comment by pllim** [comment]
Need a change log. Thanks!

**Comment by rosteen** [comment]
> Need a change log. Thanks!

I know! I thought we agreed that can wait until after approvals in case the PR drags on with lots of changes and conflicts in the changelog happen 😅

**Comment by rosteen** [comment]
The NIRCam test data download link is [here ](https://stsci.box.com/shared/static/itk7pav073nubwn58pig002m9796qzpw.zip). Working on uploading the level 2 NIRSpec data I've been using to Box as well.

**Comment by rosteen** [comment]
A sample NIRSpec level 2 dataset for testing is now [on Box](https://stsci.box.com/shared/static/mytqf082lpbfia7wlwjq6p1h5cggd9h6.zip) as well.

File: jdaviz/configs/mosviz/helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/helper.py:423]: A little confusing here since it is not in the signature above, but rather extracted from `kwargs` but I guess we can clean up the docstring in a future PR if needed.
```suggestion
        directory : str, optional
```

File: jdaviz/configs/mosviz/helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/helper.py:427]: Might as well use the actual case from code?
```suggestion
        instrument : {'niriss', 'nircam', 'nirspec'}, optional
```

File: jdaviz/configs/mosviz/helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/helper.py:429]:
```suggestion
            Required if ``directory`` is specified. Value is not case sensitive.
```

File: jdaviz/configs/mosviz/helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/helper.py:446]: Actual check is already done above, so no need to repeat here.
```suggestion
            else:  # niriss or nircam
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:225]: Please comment in the code of the *intent* here. It is not clear to me why this try-except is necessary.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:478]:
```suggestion
    sp1d : bool, optional
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:479]:
```suggestion
        Indicates if the input ``data_obj`` is a 1D rather than 2D spectrum.
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:480]:
```suggestion
    repeat : int, optional
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:501]: Easier to read if you add whitespace around the `*`. Same comment applies to all similar occurrences.

```suggestion
            filters_gratings = [(f+'/'+g) for f, g in zip(filters, gratings)] * repeat
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:553]:
```suggestion
    allow_duplicates : bool
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:610]:
```suggestion
            if fp.name in ("cutouts", "mosviz_cutouts", "images"):
```

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:735]: I never liked printing stuff out to the notebook. While you are at this, should we move this to the snackbar info?

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:749]: Is this TODO for this PR or future work?

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:845]:
```suggestion
                    filtered_hdul = fits.HDUList([hdu for hdu in temp if (
                        (hdu.name in ('PRIMARY', 'ASDF')) or
                        (hdu.header.get('SOURCEID', None) in cat_id_dict.keys()))])
```

File: jdaviz/configs/mosviz/tests/test_parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_parsers.py:92]: Why not 100 like the other one?
```suggestion
    fn = download_file(test_data, cache=True, timeout=100)
```

**Review by pllim — COMMENTED** [review]
Just a general review.

File: jdaviz/configs/mosviz/helper.py
Comment (rosteen) [reply jdaviz/configs/mosviz/helper.py:620]: Correct.

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/plugins/parsers.py:251]: If multiple objects were loaded from the same file, we need to know how many times to repeat some information in the table data.

File: jdaviz/configs/mosviz/tests/test_parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/tests/test_parsers.py:92]: Not sure why I made it different, honestly.

**Comment by javerbukh** [comment]
Was able to get NIRCAM and NIRSpec level 2 data to load, nice work!! I can approve once CI passes or if local CI passes (assuming that astropy CI bug is ongoing).

**Comment by pllim** [comment]
Nice try but change log is not astropy's fault. 😆

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/plugins/parsers.py:749]: Great question - I think I should do it here (or alternatively at least think about it and decide it's not needed). I can do that after meetings today.

**Comment by rosteen** [comment]
Docs built locally:
Screen Shot 2022-11-21 at 3 36 28 PM

**Comment by pllim** [comment]
@rosteen , if you rebase again, doc should pass.

File: docs/mosviz/import_data.rst
Comment (pllim) [inline docs/mosviz/import_data.rst:17]: Why only NIRSpec has `(levels 2 and 3)` specified? What about NIRISS and NIRCam?

File: docs/mosviz/import_data.rst
Comment (pllim) [inline docs/mosviz/import_data.rst:22]: To be consistent.
```suggestion
* ``*_x1d.fits`` : Single file containing level 2 1D spectra for all objects.
```

File: docs/mosviz/import_data.rst
Comment (pllim) [inline docs/mosviz/import_data.rst:27]: I don't think ``x2d`` exists. I cannot find it in https://jwst-pipeline.readthedocs.io/en/stable/jwst/data_products/product_types.html#stage-3-data-products

```suggestion
* ``*_x1d.fits`` : N files containing level 3 1D spectra, where N is the number of objects.
```

File: docs/mosviz/import_data.rst
Comment (pllim) [inline docs/mosviz/import_data.rst:53]: I think we want to keep this sentence but tell the user now that it would throw an error.

File: jdaviz/configs/mosviz/tests/test_parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_parsers.py:36]: It feels unnatural to `return` in a test function. Maybe this is a sign that parametrize is not suitable for this test case anymore. You don't need remote data to prove that the function will throw an exception, right? Theoretically, you could give a dummy file name/object and it would still throw the same exception?

Anyways, if we want to chuck this in the technical debt bin, we can refactor in the future and leave this as-is.

File: docs/mosviz/import_data.rst
Comment (rosteen) [reply docs/mosviz/import_data.rst:17]: NIRISS and NIRCam only have level 2 available from what I've been told, there is no level 3.

File: docs/mosviz/import_data.rst
Comment (rosteen) [reply docs/mosviz/import_data.rst:22]: Whoops, good catch.

**Review by javerbukh — APPROVED** [review]

File: jdaviz/configs/mosviz/tests/test_parsers.py
Comment (rosteen) [reply jdaviz/configs/mosviz/tests/test_parsers.py:36]: Into the bin!

**Comment by pllim** [comment]
If you rebase, CI should be all green. FYI.

**Review by pllim — APPROVED** [review]
Code looks like code. Thanks!

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (kecnry) [reply jdaviz/configs/mosviz/plugins/parsers.py:253]: This is being treated as an error for the snackbar because of https://github.com/spacetelescope/jdaviz/blob/5232ca0af9bd0425eb07647046babf8ef09b9a75/jdaviz/app.py#L530-L538

Any ideas what we should change to prevent tha twhile keeping the information needed for the mos table?

Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:253]: Wut.... I didn't even know about this behavior. Should we just remove it? Parser can emit their own messages already.

Comment (kecnry) [reply jdaviz/configs/mosviz/plugins/parsers.py:253]: If no parser currently relies on returning something other than None to provide the error message, then sure!  Otherwise I think we could easily migrate anything to emitting their own snackbar error instead of making this assumption.

Comment (kecnry) [reply jdaviz/configs/mosviz/plugins/parsers.py:253]: Created #1888 to get this in a follow-up.  Until then this just results in an error snackbar message with a number in it, which might cause a bit of confusion, but isn't critical.

**Pull request merged**
